### PR TITLE
WIP: support for opsworks for chef automate

### LINKED ISSUE #403: [Feature Request] OpsWorks for Chef Automate Support
_This issue was originally opened by @eedwardsdisco as hashicorp/terraform#10478. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Just announced at re:Invent 2016
https://www.chef.io/opsworks/

OpsWorks now supports Chef Automate (a managed chef server... woo!)

This is a feature request to add support for the new opsworks service to Terraform :)

### PR DESCRIPTION
**Pull request by kitchen**
Fixes #403

Changes proposed in this pull request:

* Add support for provisioning OpsWorks for Chef Automate instances
* Tests and documentation for above

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSAvailabilityZones'

WIP WIP WIP WIP WIP
...
```

### CODE REVIEW

File: aws/resource_aws_opsworks_chef.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+			"chef_delivery_admin_password": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Default:   nil,
+				Sensitive: true,
+			},
+
+			// TODO:
+			// default and only valid. I'll leave it configurable for future proofing but
+			// validation failure, maybe?
```
Comment (kitchen) [inline aws/resource_aws_opsworks_chef.go]: These are required parameters to the API call, but they only have one valid value at the moment.

Should I remove these from the resource and hardcode the values, leave them on the resource but enforce them in validation, or leave them on the resource, make them optional with these values as the default, and let the API handle validation? Or something else entirely?

File: aws/resource_aws_opsworks_chef.go [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+			"preferred_maintenance_window": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			// TODO:
+			// if you don't specify this, it'll create one. I'm of the opinion that users should
+			// create it themselves. Or maybe some sort of wrapper module? I dunno.
+			// if it's optional, will they be computed? if so, is that something we can express here?
```
Comment (kitchen) [inline aws/resource_aws_opsworks_chef.go]: Here's another "magic happens" thing. If you don't specify security groups, it'll make one for you. Is this something we want to allow or should we require folks to specify the security groups explicitly? I'm leaning on the side of explicitly.

File: aws/resource_aws_opsworks_chef.go [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+			// The IDs of subnets in which to launch the server EC2 instance.
+			//
+			// Amazon EC2-Classic customers: This field is required. All servers must run
+			// within a VPC. The VPC must have "Auto Assign Public IP" enabled.
+			//
+			// EC2-VPC customers: This field is optional. If you do not specify subnet IDs,
+			// your EC2 instances are created in a default subnet that is selected by Amazon
+			// EC2. If you specify subnet IDs, the VPC must have "Auto Assign Public IP"
+			// enabled.
+			// if it's optional, will they be computed? if so, is that something we can express here?
```
Comment (kitchen) [inline aws/resource_aws_opsworks_chef.go]: All sorts of :wat: here.

I'm thinking we should force people to specify subnet IDs. But that's just me.

File: aws/resource_aws_opsworks_chef.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsOpsworksChefValidate(d *schema.ResourceData) error {
+	// TODO: validation function
```
Comment (kitchen) [inline aws/resource_aws_opsworks_chef.go]: For validation, how much should I replicate the validation that amazon is going to do on the backend here?

The maintenance and backup windows have a simple format that could be validated, and then there's the chef admin password. Oh, and the chef_pivotal_key. How much should I validate before just leaving it to the API to explode when the input is wrong?

File: aws/resource_aws_opsworks_chef.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+	log.Printf("[DEBUG] Creating OpsWorks Chef server: %s", req)
+
+	var resp *opsworkscm.CreateServerOutput
+	err = resource.Retry(20*time.Minute, func() *resource.RetryError {
+		var cerr error
+		resp, cerr = client.CreateServer(req)
+		if cerr != nil {
+			if opserr, ok := cerr.(awserr.Error); ok {
+				// TODO: does this also happen with this resource?
```
Comment (kitchen) [inline aws/resource_aws_opsworks_chef.go]: Should I err on the side of "we can add this back later if it turns out it's a problem" or should I try to recreate the circumstances described to see if it needs to stay?

File: aws/resource_aws_opsworks_chef.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+func resourceAwsOpsworksChefUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AWSClient).opsworkscmconn
+
+	err := resourceAwsOpsworksChefValidate(d)
+	if err != nil {
+		return err
+	}
+
+	// TODO: these are the only values that are allowed to be changed
+	// according to the API, is there a way to express this in the provider somehow?
```
Comment (kitchen) [inline aws/resource_aws_opsworks_chef.go]: question here: only some of the values are able to be updated. The others are write-once. How to handle?

There are actually 2 more that can be updated: the `chef_pivotal_key` and `chef_delivery_admin_password`, but those are actually set using a different API call (easy enough to handle, though) but the question there is how do we know if those values need to change. Is there a way we can store the output of the createserver in a hashed form somehow so we can compare the inputs to what we last knew the values were on the server and update them? Or something else?

Comment (kitchen) [reply aws/resource_aws_opsworks_chef.go]: Seems like `ForceNew` on the attribute is the Schema is what I'm after for the first part of that question.

Comment (kitchen) [reply aws/resource_aws_opsworks_chef.go]: followup question on the ForceNew: will that propagate to any node associations that exist? Also, I should probably create a resource for those.

https://docs.aws.amazon.com/sdk-for-go/api/service/opsworkscm/#OpsWorksCM.AssociateNode

Comment (kitchen) [reply aws/resource_aws_opsworks_chef.go]: hmm. Now that I think of it, the AssociateNode is probably going to be done outside of terraform scope, which means terraform wouldn't have the info needed to be able to recreate the associations. So a ForceNew would be a pretty destructive thing.

I'm curious how this is even handled on AWS side. If instance type is one of the forcenew params, then I'm assuming they're not able to roll the backend chef machine(s) to new sizes, which makes me wonder if they even have multiples or if it's HA or whatever, what happens when one dies? I'm probably reading too much into that part of things, but I might contact amazon support and ask for more details.

Comment (kitchen) [reply aws/resource_aws_opsworks_chef.go]: Ok, I filed a support request with amazon to see what they have to say about this. For now, I'll leave ForceNew on, but I actually think forcing new here would be a really bad idea, so I'm hoping there are better ways to handle that situation :) I'm open to any suggestions.

File: aws/resource_aws_opsworks_chef.go (hunk 4 above)
Comment (kitchen) [reply aws/resource_aws_opsworks_chef.go]: ok so window format is already a thing that is validated elsewhere so I'll validate it too. I wanted to wrap the existing validators but wasn't really sure how to do it so I made another that did the needful. Definitely open to any suggestions for improvement :)

Comment (kitchen) [reply aws/resource_aws_opsworks_chef.go]: and nevermind, I found https://github.com/hashicorp/terraform/blob/a95d97f0665f905411d4e6ab400136c50539d6c6/helper/validation/validation.go#L33-L47 and am incorporating :)

Comment (kitchen) [reply aws/resource_aws_opsworks_chef.go]: annnnd nevermind, that's in a newer version of terraform than this package is pulling in (I think, I'm not super familiar with golang development tbh). I'll use the one in this provider: https://github.com/terraform-providers/terraform-provider-aws/blob/85e58f8cf3c79621d32db94f657e8a258ee44fd3/aws/validators.go#L27-L42

File: aws/resource_aws_opsworks_chef.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+				Optional: true,
+				Default:  false,
+			},
+
+			"backup_retention_count": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1,
+				// https://docs.aws.amazon.com/opsworks/latest/userguide/opscm-chef-backup.html
+				// I could swear I found this 1-30 limitation elsewhere, but I can't seem to find it now other than there
```
Comment (atward) [inline aws/resource_aws_opsworks_chef.go]: Found docs [here](https://docs.aws.amazon.com/opsworks/latest/userguide/gettingstarted-opscm-create.html)
> A maximum of 30 backups are kept; when the maximum is reached, AWS OpsWorks for Chef Automate deletes the oldest backups to make room for new ones.
It's also on the console wizard

**Comment by tylermauthe** [comment]
Hello! Just curious if this PR is still being worked on; it would be a great benefit to me. 🎉 

Is there anything I can do to assist -- or just shut up and wait 🤐?

Thanks!

**Comment by kitchen** [comment]
It is not currently being worked on, by me anyways, sorry :(

I don’t remember quite where I left off with it but I imagine it’s missing docs, some fixes on the comments I had and had gotten replies to, and maybe a few other things. I will be out of town for 2 weeks starting in a few days so I won’t have much of an opportunity to work in it for a while.

**Comment by ewbankkit** [comment]
You can now add tags to an AWS OpsWorks for Chef Automate server or an AWS OpsWorks for Puppet Enterprise master, or to server backups.

Requires [AWS SDK v1.26.5](https://github.com/aws/aws-sdk-go/releases/tag/v1.26.5):
* https://github.com/terraform-providers/terraform-provider-aws/pull/11353

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by ewbankkit** [comment]
Closing as stale due to inactivity.

**Pull request closed without merging**